[Sema] Fix several redeclaration checking bugs

hamishknight opened the pull request:
Currently we only give subscripts and var decls custom overload types if they're in generic extensions. However, because we give them no custom overload type in any other case, we don't detect for example a conflict with a previous declaration in the body of the extended type – leading to accepting invalid and crashing in SILGen.

This commit changes the overload type logic such that properties and subscripts are always given custom overload types, which is determined by:

- The interface type of the decl (for subscripts only; as variables cannot be overloaded by type)
- The 'self' type of the context, if any
- The generic signature of the context, if any

Additionally, this commit adds a new `swift::conflicting` overload to ensure that different declarations always conflict even if their overload types are different.

Resolves [SR-7249](https://bugs.swift.org/browse/SR-7249), [SR-7250](https://bugs.swift.org/browse/SR-7250) & [SR-7251](https://bugs.swift.org/browse/SR-7251).

hamishknight committed:
[Sema] Fix several redeclaration checking bugs

Currently we only give subscripts and var decls custom overload types if they're in generic extensions. However, because we give them no custom overload type in any other case, we don't detect for example a conflict with a previous declaration in the body of the extended type.

This commit changes the overload type logic such that properties and subscripts are always given custom overload types, which is determined by:

- The interface type of the decl (for subscripts only; as variables cannot be overloaded by type)
- The 'self' type of the context, if any
- The generic signature of the context, if any

Additionally, this commit adds a new `swift::conflicting` overload to ensure that different declarations always conflict even if their overload types are different.

Resolves SR-7249, SR-7250 & SR-7251.

slavapestov commented on include/swift/AST/Decl.h:
```diff
@@ -211,8 +211,11 @@ struct OverloadSignature {
   /// Whether this is an instance member.
   bool IsInstanceMember = false;
```
These could be bitfields for efficiency (I know this isn't part of your change but it might be easy to change it)

hamishknight replied:
Good idea, thanks

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -1685,13 +1685,35 @@ bool swift::conflicting(const OverloadSignature& sig1,
+                        bool skipProtocolExtensionCheck) {
+  // If the signatures don't conflict to begin with, we're done.
+  if (!conflicting(sig1, sig2, skipProtocolExtensionCheck))
+      return false;
```
Nitpick: indent

hamishknight replied:
Oops, thanks

slavapestov reviewed (commented):
I think @DougGregor or @jrose-apple should look at this in more detail

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -1824,44 +1847,53 @@ CanType ValueDecl::getOverloadSignatureType() const {
+    auto genericCtx = parentDecl->getAsGenericContext();
+
+    // If we're in a context with a generic signature, return a generic function
+    // type with that signature. Otherwise, return a non-generic type.
```
What if the subscript itself is generic? I know that didn't work before either, but there should at least be a FIXME if you're going to redo the code without fixing it.

hamishknight replied:
@jrose-apple Looking into it, I think we already correctly handle generic subscripts in this PR. The generic signature just gets nested in the type when we curry it with 'self'.

For example, for this subscript:

```swift
struct Foo<U> {
  subscript<T>(_ t: T) -> U {
    fatalError()
  }
}
```

Its overload type is:
```
(generic_function_type escaping
  (input=paren_type
    (bound_generic_struct_type decl=main.(file).Foo
      (generic_type_param_type depth=0 index=0)))
  (output=generic_function_type escaping
    (input=paren_type
      (generic_type_param_type depth=1 index=0))
    (output=generic_type_param_type depth=0 index=0)
    ( generic_sig=<τ_0_0, τ_1_0>))
  ( generic_sig=<τ_0_0>))
```

I think this is okay for the purposes for an overload type. I'll go ahead and add some tests for generic subscripts though.

jrose-apple replied:
I was pretty sure it still has to be a GenericFunctionType, though, which is something that *wouldn't* happen if the subscript is generic but the context is not.

(I'll reply to the other comments later.)

hamishknight replied:
@jrose-apple Ah okay, that makes sense. So presumably we need to lift the subscript's generic signature up to the curried function's generic signature.

Actually [just found an existing function](https://github.com/apple/swift/blob/d8b355dbaebd157ba75e19fbc6ace87ef88e3e98/lib/Sema/CSRanking.cpp#L257) that curries a type with a given context's 'self', so can just reuse that for this (it doesn't use a MetatypeType for a static member, but we didn't really need that to begin with as static members don't have conflicting signatures with instance members).

So for this subscript:
```swift
struct Foo {
  subscript<T>(_ t: T) -> T {
    fatalError()
  }
}
```
its overload type is now:
```
(generic_function_type escaping
  (input=paren_type
    (struct_type decl=main.(file).Foo))
  (output=function_type escaping
    (input=paren_type
      (generic_type_param_type depth=0 index=0))
    (output=generic_type_param_type depth=0 index=0))
  ( generic_sig=<τ_0_0>))
```

which matches up with the overload type for a similar method decl, e.g:
```swift
struct Foo {
  func foo<T>(_ t: T) -> T {
    fatalError()
  }
}
```

I've gone ahead and pushed the revisions for your feedback – thanks for your patience!

jrose-apple commented on test/decl/overload.swift:
```diff
@@ -149,8 +149,15 @@ func ovl_generic3<T : P2>(_ x: T) {} // OK
 struct Y { }
 struct Z { 
-  var a : X, // expected-note{{previously declared here}}
-  a : Y // expected-error{{invalid redeclaration of 'a'}}
```
Please don't remove this example; it may be relevant that two things are in the same pattern-binding. You can make a new struct instead if you want to test the other case.

jrose-apple commented on test/decl/overload.swift:
```diff
@@ -300,6 +343,26 @@ enum EnumWithMutating {
+  associatedtype u
+  associatedtype v
+
+  // instance requirements are fine.
```
Nitpick: please capitalize starts of sentences in comments.

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -1685,13 +1685,35 @@ bool swift::conflicting(const OverloadSignature& sig1,
+    return false;
+
+  // Variables always conflict with non-variables with the same signature.
+  // (e.g variables with zero argument functions, variables with type
```
Variables do not conflict with zero-argument functions (and you've implemented them to not conflict, so that's okay).

jrose-apple replied:
Why are these new checks here and not inside the original `conflicting`?

hamishknight replied:
@jrose-apple Variables do indeed conflict with zero argument functions (we have `mixed_redecl6` in the `overload.swift` test for that), e.g:

```
struct Foo {
  var i: Int { return 0 }
  func i() {} // error: Invalid redeclaration of 'i()'
}
```
---
These new checks are in this function as they're only relevant when comparing the overload types (they prevent us from saying that two decls don't conflict because the overload types differ).

To be honest, I'm not too mad on having this logic separated across three functions (the two `conflicting` overloads, and the logic that computes the overload types), but I couldn't really see a better way to do it given the two current use cases of the logic. Any ideas on how to better structure things are more than welcome :)

jrose-apple replied:
> Variables do indeed conflict with zero argument functions – we have `mixed_redecl6` in the `overload.swift` test for that

Aargh, right. You're not allowed to declare them, even though you can end up with that effect by importing other modules.

jrose-apple commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -957,8 +957,7 @@ static void checkRedeclaration(TypeChecker &tc, ValueDecl *current) {
     // If there is another conflict, complain.
-    if (currentSigType == otherSigType ||
-        currentSig.Name.isCompoundName() != otherSig.Name.isCompoundName()) {
+    if (conflicting(currentSig, currentSigType, otherSig, otherSigType)) {
```
This is going to redo the check from earlier. Is it possible to not do that?

hamishknight replied:
@jrose-apple I agree it's not ideal. We could pass a `bool` flag, something like `areSignaturesConflicting` – at the time I didn't really feel it was worth it given the logic being repeated was fairly lightweight, but am happy to do so if you feel it would be worthwhile.

jrose-apple reviewed (commented):
Thanks for fixing these! I'm a little suspect of some of the changes but I think the overall direction makes sense (in particular the important part where properties and subscripts now get sensible types).

hamishknight committed:
[Sema] Revise #15412 in response to feedback

- Reuse existing logic to curry the signature type with the 'self' of the context (in addition we no longer use a MetatypeType for the 'self' of a static member as they don't have conflicting signatures with instance members anyway)

- Limit the fix for SR-7251 to Swift 5 mode

- Add tests for generic subscripts

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -1685,13 +1685,45 @@ bool swift::conflicting(const OverloadSignature& sig1,
+    // Prior to Swift 5, we permitted redeclarations of variables as different
+    // declarations if the variable was declared in an extension of a generic
+    // type. Make sure we maintain this behaviour in versions < 5.
+    if (!ctx.LangOpts.EffectiveLanguageVersion.isVersionAtLeast(5))
```
We have this as a convenience predicate on ASTContext, `isSwiftVersionAtLeast`.

jrose-apple commented on lib/AST/LookupVisibleDecls.cpp:
```diff
@@ -805,10 +805,10 @@ class OverrideFilteringConsumer : public VisibleDeclConsumer {
-          (FoundSignatureType == OtherSignatureType ||
-           FoundSignature.Name.isCompoundName() !=
-             OtherSignature.Name.isCompoundName())) {
+      if (conflicting(VD->getDeclContext()->getASTContext(),
```
No need to jump up to the DeclContext here; Decls also have a `getASTContext()`. (But it'd be fastest to get it from the module `M`, which always has a direct reference.)

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -1685,13 +1685,45 @@ bool swift::conflicting(const OverloadSignature& sig1,
+                        bool skipProtocolExtensionCheck) {
+  // If the signatures don't conflict to begin with, we're done.
+  if (!conflicting(sig1, sig2, skipProtocolExtensionCheck))
+    return false;
```
I still don't know about this. It might be simpler to just have a second function that says "given that the signatures overlap, do the types conflict?" and makes this into an assertion instead of a check. I'm going to defer to @DougGregor on this one (which will slow down the review, sorry).

jrose-apple commented:
The other thing we could consider doing is *warning* in pre-Swift-5 modes that what they're doing is probably not a good idea.

hamishknight committed:
[Sema] Revise #15412 in response to feedback (2)

- Emit a warning in Swift 4 for performing a redeclaration that will be illegal in Swift 5

- Minor fixes

hamishknight commented:
@jrose-apple Good idea, thanks, I've gone ahead and made those changes (keeping the second `swift::conflicting` function chaining to the first for now though, will see what Doug Gregor thinks).

For the warning, I used an out-parameter instead of a return with 3 values in order to avoid burdening the caller in the case where they just want to know if there's a conflict between two decls in the current Swift version (which IMO should be the common case). Let me know what you think about that.

Thanks again for the feedback!

DougGregor commented on lib/AST/Decl.cpp:
```diff
@@ -1685,13 +1685,51 @@ bool swift::conflicting(const OverloadSignature& sig1,
+      if ((sig1.IsVariable && sig1.InExtensionOfGenericType) ||
+          (sig2.IsVariable && sig2.InExtensionOfGenericType)) {
+        if (wouldConflictInSwift5)
+          *wouldConflictInSwift5 = true;
```
I like this staging via a tri-state a lot, thank you!

DougGregor commented on lib/AST/Decl.cpp:
```diff
@@ -1824,44 +1867,29 @@ CanType ValueDecl::getOverloadSignatureType() const {
+    // is unique across different contexts, such as between a protocol extension
+    // and struct decl.
+    return defaultSignatureType->addCurriedSelfType(getDeclContext())
+                               ->getCanonicalType();
```
Very nice cleanup here, thank you.

DougGregor commented on include/swift/AST/Decl.h:
```diff
@@ -206,22 +206,56 @@ struct OverloadSignature {
+/// \param skipProtocolExtensionCheck If \c true, members of protocol extensions
+///        will be allowed to conflict with members of protocol declarations.
 bool conflicting(const OverloadSignature& sig1, const OverloadSignature& sig2,
                  bool skipProtocolExtensionCheck = false);
```
Can we make this previous form of `conflicting` static? It looks like only the new `conflicting` calls it.

hamishknight replied:
@DougGregor We do still call the original `conflicting` function [here](https://github.com/apple/swift/pull/15412/files#diff-32cb5387532d533d00872d725748a4d9L921) in order to initially discount decls with different signatures (before we then go to calculate their overload type) – which means that we'll end up calling the first `conflicting` twice if the signatures conflict.

Jordan raised his concerns over this redundant additional calling [here](https://github.com/apple/swift/pull/15412#discussion_r176475034) and [here](https://github.com/apple/swift/pull/15412#discussion_r176786739) – I wasn't too concerned given the logic is fairly lightweight; but it would be nice to refactor in such a way where that's not the case. Thoughts?

(I notice that `overload_swift4.swift` is failing on Linux; I'll look into that...)

DougGregor replied:
Ah, I'd missed the remaining call to `conflicting`. I'm not too concerned about the extra calls, because this is a fairly quick check.

DougGregor reviewed (approved):
This looks fantastic! I'd love to get this merged now, and my one (minor) comment could be addressed in a follow-up commit.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

hamishknight committed:
[Sema] Avoid emitting more than one warning for a redeclaration that will be illegal in Swift 5

hamishknight commented:
Apologies, there was an issue where we could emit more than one redeclaration deprecation warning per decl which I somehow didn't catch when testing locally. We don't run into the same problem for re-declaration errors as we mark those decls as invalid, which prevents us from checking them again.

I've gone ahead and pushed a fix for that.

DougGregor commented:
@hamishknight , I wonder if your change fixes this example we just tripper over elsewhere:

```
struct A<Element> {
	var a: String {
		return "a1"
	}
}

extension A {
	var a: String {
		return "a2"
	}
}
```

hamishknight commented:
@DougGregor Yes, that's [SR-7249](https://bugs.swift.org/browse/SR-7249) and is fixed by this PR (in that we now emit a redeclaration error for it). We have a test case for it [here](https://github.com/apple/swift/pull/15412/files#diff-77d426ccfae20f18ff28083b5bfc6d0eR442).

DougGregor commented:
Ah hah, great!

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
drat, something conflicted before merge.

DougGregor committed:
Merge branch 'master' into redeclaration-fixes

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

hamishknight commented:
Thanks for resolving the merge conflict @DougGregor! Looks like this is good to go now.

DougGregor commented:
Thank you! Want to do the honors of resolving the JIRAs, @hamishknight ?

aschwaighofer commented:
This appears to have broken ReactiveCocoa in the source compatibility suite

aschwaighofer commented:
https://ci.swift.org/job/swift-master-source-compat-suite/1527/console

jrose-apple commented:
Failure logs: https://ci.swift.org/job/swift-master-source-compat-suite/1527/artifact/swift-source-compat-suite/

```
/Users/buildnode/jenkins/workspace-private/swift-master-source-compat-suite/project_cache/ReactiveCocoa/ReactiveCocoa/NSObject+Intercepting.swift:389:19: error: ambiguous use of 'numberOfArguments'
        let count = UInt(methodSignature.numberOfArguments!)
                         ^
/Users/buildnode/jenkins/workspace-private/swift-master-source-compat-suite/project_cache/ReactiveCocoa/ReactiveCocoa/ObjC+Messages.swift:38:6: note: found this candidate
        var numberOfArguments: UInt { get }
            ^
Foundation.NSMethodSignature:7:14: note: found this candidate
    open var numberOfArguments: Int { get }
             ^
```

Maybe we were filtering this out before? Or maybe they meant to use `as UInt`?

hamishknight commented:
Here's a minimal example of the issue:

```swift
import Foundation

@objc protocol P {
  var i: UInt { get }
}

class C : NSObject {
  @objc var i: Int {
    return 0
  }
}

let x = (C() as AnyObject).i // compiles before patch, "ambiguous use of 'i'" error after
```
Presumably they meant to make the protocol requirement match the interface for `NSMethodSignature` (i.e make it `var numberOfArguments: Int { get }`) – that would fix the issue.

I'm not too familiar with the workings of the "ambiguous use of" diagnostic, but before the patch we would have given both `i` in the protocol and the class decl the same overload type (`CanType()`), whereas now they have different overload types.

So presumably we now see them as two different overloads and diagnose that there's not enough context to tell them apart (the correct behaviour IMO), rather than seeing them as the same overload and favouring the `Int` overload over the `UInt` one.

What course of action should we take here? Do we need to find a way to downgrade the ambiguity error to a warning until Swift 5 mode?

aschwaighofer commented:
@clackary Has XFAILed ReactiveCocoa for now. https://bugs.swift.org/browse/SR-7299

aschwaighofer commented:
@jrose-apple @DougGregor @hamishknight : In the short term what is the way forward? To do you intend to update ReactiveCocoa (or talk to the maintainers to do so) or should we revert until we have a fix: new behavior guarded by swift mode? I think we are trying hard not to have source compat projects disabled for too long ...

jrose-apple commented:
Doug and I are leaning towards asking the ReactiveCocoa folks for an update. The old code would have been picking a property basically at random (whatever it saw first), and we can't think of a sensible way to downgrade this one to a warning.

hamishknight commented:
@jrose-apple Do you know why `NSMethodSignature`'s `numberOfArguments` is being imported into Swift as an `Int` [despite being documented as being](https://developer.apple.com/documentation/foundation/nsmethodsignature/1519662-numberofarguments?language=objc) an `NSUInteger`, and therefore should (AFAIK) be imported as a `UInt`?

Just trying to figure out why the issue is only reproducible for them on iOS and tvOS, but not macOS.

hamishknight commented:
@andersio has opened a PR here to fix things for ReactiveCocoa: https://github.com/ReactiveCocoa/ReactiveCocoa/pull/3580
